src/sqlite_db.cpp

```diff
@@ -50,6 +50,243 @@
 
 namespace miopen {
 
+class SQLite::impl
+{
+    struct SQLiteCloser
+    {
+        void operator()(sqlite3* ptr)
+        {
+            std::string filename_(sqlite3_db_filename(ptr, "main"));
+            SQLite::impl::Retry([&]() { return sqlite3_close(ptr); }, filename_);
+        }
+    };
+
+    public:
+    impl(const std::string& filename_, bool is_system)
+    {
+        sqlite3* ptr_tmp;
+        int rc = 0;
+        if(is_system)
+            rc = sqlite3_open_v2(filename_.c_str(), &ptr_tmp, SQLITE_OPEN_READONLY, nullptr);
+        else
+            rc = sqlite3_open_v2(
+                filename_.c_str(), &ptr_tmp, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, nullptr);
+        ptrDb   = sqlite3_ptr{ptr_tmp};
+        isValid = (rc == 0);
+    }
+    bool Exec(const std::string& query) const
```

Comment:
The methods should be implemented in the main class, and not in the impl class. This requires too much boilerplate to add a method.

The impl class should just store the sqlite pointer.

Reply:
Moved all the methods out of the impl class ( for SQLite and Statement)